Memory Stream - File Stream (win/lin) - HTTP Request / Response - Updates

vhvb1989 opened the pull request:
Adding:
-  File Stream for Linux and Windows.  Customer is responsible for opening the file to get a handle. Then we use that handle to construct the BodyStream for the File.  See the basic example on how to use it on sdk/Samples/azure_core_with_curl_bodyBuffer.cpp  (it is currently only running a File Stream test)

- Memory Stream
   - Removing the memory stream that holds a vector within. We don't need to have such a stream.
   - Keeping only Memory stream that works with a ptr and size
   - The programming model needs to be adapted to avoid using "constructRequest / CreateRequest" methods where the body content is created and expected to be moved to the memory stream. Instead, create the body content in the same method where the request will be sent to keep the body content in stack. (See the change - examples I made to blob_rest_client.hpp)
   
  - Adding static methods for reading memory stream.
```cpp
// Will loop reading from body until buffer is filled up to count bytes OR until the end of body content
static int64_t ReadToCount(Context& context, BodyStream& body, uint8_t* buffer, int64_t count);

// Will loop reading and inserting bytes into an std::vector<> until the end of body content
static std::unique_ptr<std::vector<uint8_t>> ReadToEnd(Context& context, BodyStream& body);
```

- HTTP Request is updated in this PR to avoid owning the body stream. This is a proposal to make Request have a ref BodyStream& so the customer/convenience layer can keep the body stream ownership instead of moving it to the request. Allowing a customer to reuse a unique_ptr<BodyStream>.


>Note: Storage cmake project is currently disabled because it need to be updated on top of Core Changes.  Please have it enabled by updating CMakeLists.txt upon fixing breaking changes.

>Note2: I am still working on adjusting the internal implementation of the libcurl transport to be fully function with this latest changes.  However, the signatures won't change, so please start fixing breaking changes to Storage client

Fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/198
Fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/195
Fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/194
Fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/96

vhvb1989 commented on CMakeLists.txt:
```diff
@@ -43,4 +43,4 @@ include(${CMAKE_SOURCE_DIR}/cmake-modules/doxygen_common.cmake)
 add_subdirectory(sdk/core/azure-core)
 add_subdirectory(sdk/samples/http_client/curl) # will work only if BUILD_CURL_TRANSPORT=ON
-add_subdirectory(sdk/storage)
+#add_subdirectory(sdk/storage)
```
uncomment this to re-enable storage cmake proj

JinmingHu-MSFT replied:
I think it's a very good practice to disable storage build when you introduce a breaking change in Azure Core.

Then why did you also change storage code?

vhvb1989 commented on sdk/samples/http_client/curl/src/azure_core_with_curl_bodyBuffer.cpp:
```diff
@@ -82,23 +73,67 @@ int main()
+#ifdef Posix
+void doFileRequest(Context context, HttpPipeline& pipeline)
+{
+  (void)pipeline;
+  string host("https://httpbin.org/put");
+  cout << "Creating a File request to" << endl << "Host: " << host << endl;
+
+  int fd = open("/home/vagrant/workspace/a", O_RDONLY);
+  auto requestBodyStream = std::make_unique<FileBodyStream>(fd, 0, 10);
+
+  auto body = Http::BodyStream::ReadToEnd(context, *requestBodyStream);
+  cout << body->data() << endl << body->size() << endl;
+
+  close(fd);
```
linux File Stream Example

vhvb1989 commented on sdk/samples/http_client/curl/src/azure_core_with_curl_bodyBuffer.cpp:
```diff
@@ -82,23 +73,67 @@ int main()
+#ifdef Windows
+void doFileRequest(Context context, HttpPipeline& pipeline)
+{
+  (void)pipeline;
+  string host("https://httpbin.org/put");
+  cout << "Creating a File request to" << endl << "Host: " << host << endl;
+
+  // NOTE: To run the sample: Create folder 'home' on main hard drive (like C:/) and then add a file
+  // `a` in there
+  //
+  HANDLE hFile = CreateFile(
+      "/home/a",
+      GENERIC_READ,
+      FILE_SHARE_READ,
+      NULL,
+      OPEN_EXISTING,
+      FILE_FLAG_SEQUENTIAL_SCAN,
+      NULL);
+  auto requestBodyStream = std::make_unique<FileBodyStream>(hFile, 20, 200);
+
+  auto body = Http::BodyStream::ReadToEnd(context, *requestBodyStream);
+  cout << body->data() << endl << body->size() << endl;
+
+  CloseHandle(hFile);
```
Windows File Stream Example

vhvb1989 commented on sdk/storage/inc/common/memory_stream.hpp:
```diff
@@ -32,7 +30,7 @@ namespace Azure { namespace Storage {
-  inline std::unique_ptr<MemoryStream> CreateMemoryStream(const uint8_t* data, std::size_t length)
+  inline std::unique_ptr<MemoryStream> CreateMemoryStream(const uint8_t* data, int64_t length)
   {
     return std::make_unique<MemoryStream>(data, length);
```
Can we remove this MemoryStream ?   Use only the one from Core now?

JinmingHu-MSFT replied:
Yes, I see you've changed the implementation of `Azure::Core::MemoryStream`, it doesn't make a copy now.

JinmingHu-MSFT replied:
done

vhvb1989 commented on sdk/storage/src/blobs/block_blob_client.cpp:
```diff
@@ -58,11 +58,11 @@ namespace Azure { namespace Storage { namespace Blobs {
   {
     BlobRestClient::BlockBlob::UploadOptions protocolLayerOptions;
-    protocolLayerOptions.BodyStream = std::move(content);
+    // protocolLayerOptions.BodyStream = std::move(content);
```
We also want to move body content out of options.
BodyStream is a required parameter for HttpRequest, just like url.
So it should not be in operation options.

Confirmed by @jeff  and @RickWinter

JinmingHu-MSFT replied:
Blob service changed.

katmsft replied:
I am still in doubt why body stream should be required for **ALL** requests. A request body should never be mandatory even for PUT requests.

JinmingHu-MSFT replied:
I agree that request body shouldn't be mandatory for GET/HEAD/DELETE requests. But I think it's mandatory for PUT/POST requests.

JeffreyRichter replied:
HTTP allows request bodies for ALL HTTP methods and so the request supports this. Also, the request object doesn't even know the exact method that's being used (it's passed as a parameter). So, since every method CAN have a body and because the body is a separate logical entity of any arbitrary HTTP request the body is an explicitly specified parameter. We added a NullBodyStream which can be passed for HTTP requests that want to have no body.

vhvb1989 commented on sdk/storage/test/test_base.hpp:
```diff
@@ -29,9 +29,14 @@ namespace Azure { namespace Storage { namespace Test {
 
-  std::vector<uint8_t> ReadBodyStream(std::unique_ptr<Azure::Core::Http::BodyStream>& stream);
+  inline std::vector<uint8_t> ReadBodyStream(std::unique_ptr<Azure::Core::Http::BodyStream>& stream)
+  {
+    Azure::Core::Context context;
+    return *Azure::Core::Http::BodyStream::ReadToEnd(context, *stream);
+  }
 
-  inline std::vector<uint8_t> ReadBodyStream(std::unique_ptr<Azure::Core::Http::BodyStream>&& stream)
+  inline std::vector<uint8_t> ReadBodyStream(
+      std::unique_ptr<Azure::Core::Http::BodyStream>&& stream)
```
We would not need these ones anymore.  Update to use the BodyStream::ReadToCount(...) or BodyStream::ReadToEnd(..) functions

JinmingHu-MSFT replied:
I think so

JinmingHu-MSFT replied:
> @JinmingHu-MSFT , yes, 9 and 10 are fixed.
> 
> And the seg fault was an error in the sample test cases. I have fixed.
> 
> I was getting the hang while reading and found the root cause. I have tested Core transport implementation with the samples I have and it is working fine with the changes on this PR.
> 
> Please fix Storage Code whenever possible
> 
> About the BodyStream Reference. Core team will come to a final decision about this tomorrow. (I will update PR after that)

9 isn't fixed.

Here https://github.com/Azure/azure-sdk-for-cpp/blob/74c84147c5b8d673ecddb58926704ae92ed49ea5/sdk/core/azure-core/src/http/curl/curl.cpp#L444,

if I remove Line 444, then 9 is fixed.

I didn't look into it carefully, so please double check the logic here is correct.

JinmingHu-MSFT replied:
Now it's fixed

JinmingHu-MSFT commented on sdk/core/azure-core/inc/http/bodyStream.hpp:
```diff
@@ -0,0 +1,269 @@
+    // Keep reading until buffer is all fill out of the end of stream content is reached
+    static int64_t ReadToCount(Context& context, BodyStream& body, uint8_t* buffer, int64_t count)
+    {
+      int64_t readBytes;
```
`readBytes` isn't used out of the for loop, so maybe you want to move it into the loop

JinmingHu-MSFT commented on sdk/core/azure-core/inc/http/bodyStream.hpp:
```diff
@@ -0,0 +1,269 @@
+    MemoryBodyStream(std::vector<uint8_t> const&&) = delete;
+
+    MemoryBodyStream(std::vector<uint8_t> const& buffer)
+        : MemoryBodyStream(buffer.data(), static_cast<uint64_t>(buffer.size()))
```
you want to static_cast to int64_t? looks like a typo

JinmingHu-MSFT commented on sdk/core/azure-core/inc/http/bodyStream.hpp:
```diff
@@ -0,0 +1,269 @@
+
+      int64_t copy_length = std::min(count, (int64_t)this->m_length - this->m_offset);
+      // Copy what's left or just the count
+      std::memcpy(buffer, this->m_data + m_offset, (size_t)copy_length);
```
I remember compiler will give a warning for `std::memcpy` on Windows.

JinmingHu-MSFT commented on sdk/core/azure-core/inc/http/bodyStream.hpp:
```diff
@@ -0,0 +1,269 @@
+    {
+      context.ThrowIfCanceled();
+
+      int64_t copy_length = std::min(count, (int64_t)this->m_length - this->m_offset);
```
One conclusion from Jun. 10 meeting is

* Use C++ style casts
* Avoid C casts.

reference to the recording at around 52 min.

JinmingHu-MSFT replied:
ditto for all other casts, since I've seen these C style casts a lot in your code.

JinmingHu-MSFT commented on sdk/core/azure-core/inc/http/http.hpp:
```diff
@@ -215,7 +215,7 @@ namespace Azure { namespace Core { namespace Http {
     std::map<std::string, std::string> m_retryQueryParameters;
     // Work only with streams
-    std::unique_ptr<BodyStream> m_bodyStream;
+    BodyStream& m_bodyStream;
```
Why is this changed to reference again? It was just changed to unique_ptr several days ago.

I vaguely remember that we don't allow reference in member variables.

JinmingHu-MSFT replied:
here https://azure.github.io/azure-sdk/cpp_design.html#const-and-reference-members

vhvb1989 replied:
https://github.com/Azure/azure-sdk-for-cpp/issues/238

katmsft replied:
It would be best for us to avoid this kind of breaking changes in the future. Changing from A to B is fine, but changing from A to B to C then back to A seems to be badly planned or designed.
Also, there is no detailed information in the linked issue for the Storage team to know the reasoning behind this change.

JeffreyRichter replied:
We're experimenting with different designs and getting input from Billy/Casey. Right now, I cannot guarantee that this will not change again - in fact, we're meeting on it again today. We want to get this right and there are nuances. C++ makes it very difficult; in other languages this was trivial. Please just bear with us a while longer.

JinmingHu-MSFT commented on sdk/core/azure-core/CMakeLists.txt:
```diff
@@ -14,6 +14,13 @@ if(NOT CURL_FOUND)
   find_package(CURL ${CURL_MIN_REQUIRED_VERSION} REQUIRED)
 endif()
 
+if (MSVC)
```
What if Cygwin and MinGW?

JinmingHu-MSFT commented on sdk/core/azure-core/inc/http/bodyStream.hpp:
```diff
@@ -0,0 +1,269 @@
+    int64_t Length() const override { return this->m_length; };
+
+    // close does nothing opp
+    void Close() {}
```
```suggestion
    void Close() override {}
```

JinmingHu-MSFT commented on sdk/core/azure-core/inc/http/bodyStream.hpp:
```diff
@@ -0,0 +1,269 @@
+    int64_t Length() const override { return this->m_length; };
+
+    // close does nothing opp
+    void Close() {}
```
```suggestion
    void Close() override {}
```

JinmingHu-MSFT commented on sdk/core/azure-core/CMakeLists.txt:
```diff
@@ -14,6 +14,13 @@ if(NOT CURL_FOUND)
 endif()
 
+if (MSVC)
+  add_compile_definitions(Windows)
```
It's uncommon to define macro with lowercase letters?

JinmingHu-MSFT commented on sdk/core/azure-core/inc/http/bodyStream.hpp:
```diff
@@ -0,0 +1,269 @@
+    {
+      context.ThrowIfCanceled();
+
+      auto result = pread(
```
`pread()` may return `-1` on error.

vhvb1989 replied:
https://github.com/Azure/azure-sdk-for-cpp/issues/237

JinmingHu-MSFT commented on sdk/core/azure-core/inc/http/bodyStream.hpp:
```diff
@@ -0,0 +1,269 @@
+      o.Offset = (DWORD)(this->m_baseOffset + this->m_offset);
+      o.OffsetHigh = (DWORD)((this->m_baseOffset + this->m_offset) >> 32);
+
+      auto result = ReadFile(
```
error handling

vhvb1989 replied:
https://github.com/Azure/azure-sdk-for-cpp/issues/237

JinmingHu-MSFT commented on sdk/core/azure-core/inc/http/bodyStream.hpp:
```diff
@@ -0,0 +1,269 @@
+#pragma once
+
+#ifdef Posix
+#include <unistd.h>
+#endif
+
+#ifdef Windows
+#define WIN32_LEAN_AND_MEAN
+#define NOMINMAX
+#include <Windows.h>
+#endif // Windows
```
Nit: is it a good idea to include `unistd.h` and `Windows.h` in header file? It'll pollute the global namespace.

JeffreyRichter replied:
Obviously, we prefer not to do this. For linux we can problem get rid of it because a file descriptor is an 'int'.
For Windows, the HANDLE is the data type. Maybe we can get rid of this header file with a small hack and define the FileBodyStream ctor parameter as a uint32_t. We'll just have to add some comment/documentation.

JinmingHu-MSFT commented:
@vhvb1989 BTW, are the remaining two bugs in https://github.com/Azure/azure-sdk-for-cpp/issues/198 also fixed in this PR? Because they are still blockers for storage team.

vhvb1989 commented:
> @vhvb1989 BTW, are the remaining two bugs in #198 also fixed in this PR? Because they are still blockers for storage team.

are they the 9 and 10?   Added info about them.
Let me know if there is something else?

I am seeing a seg fault right now, I will fix it here

JinmingHu-MSFT commented:
> if there is something

9 and 11, I struckthrough the ones that have been fixed.

JinmingHu-MSFT commented:
> I am seeing a seg fault right now, I will fix it here

You don't need to fix crashes in storage code, we can take care of that. You only need to ensure there's no crash in core code.

vhvb1989 commented:
@JinmingHu-MSFT , yes, 9 and 10 are fixed.

And the seg fault was an error in the sample test cases. I have fixed.

I was getting the hang while reading and found the root cause. I have tested Core transport implementation with the samples I have and it is working fine with the changes on this PR.

Please fix Storage Code whenever possible

About the BodyStream Reference.  Core team will come to a final decision about this tomorrow.  (I will update PR after that)

JinmingHu-MSFT commented on cmake-modules/DefinePlatform.cmake:
```diff
@@ -0,0 +1,18 @@
+# Cmake variables:
+# UNIX   : is TRUE on all UNIX-like OS's, including Apple OS X and CygWin
+
+# WIN32  : is TRUE on Windows. Prior to 2.8.4 this included CygWin
+# MINGW  : is TRUE when using the MinGW compiler in Windows
+# MSYS   : is TRUE when using the MSYS developer environment in Windows
+# CYGWIN : is TRUE on Windows when using the CygWin version of cmake
+
+# APPLE  : is TRUE on Apple systems. Note this does not imply the
+#          system is Mac OS X, only that APPLE is #defined in C/C++
+#          header files.
+
+if (WIN32 OR MINGW OR MSYS OR CYGWIN)
+  add_compile_definitions(WINDOWS)
+endif ()
+if (UNIX)
+  add_compile_definitions(POSIX)
+endif ()
```
Do we really need to make it so compilcated?

I think in the code

```
#ifdef _WIN32
#include <Windows.h>
#else
#include <unistd.h>
#endif
```
should be enough.

JinmingHu-MSFT commented on sdk/core/azure-core/inc/http/http.hpp:
```diff
@@ -233,14 +233,17 @@ namespace Azure { namespace Core { namespace Http {
     // Typically used for GET with no request body.
-    Request(HttpMethod httpMethod, std::string const& url) : Request(httpMethod, url, nullptr) {}
+    Request(HttpMethod httpMethod, std::string const& url)
+        : Request(httpMethod, url, *std::make_unique<NullBodyStream>())
```
I don't think this will work, the `std::unique_ptr` will be destructed very soon, then you're referencing a deleted object.

JinmingHu-MSFT replied:
This leads to a crash issue.

vhvb1989 replied:
fixed.

JinmingHu-MSFT commented on sdk/core/azure-core/CMakeLists.txt:
```diff
@@ -22,7 +22,7 @@ add_library (
   src/http/request.cpp
   src/http/response.cpp
-  src/http/stream.cpp
+  src/http/bodyStream.cpp
```
```suggestion
  src/http/body_stream.cpp
```

JinmingHu-MSFT replied:
we have file naming rules. Other files are using snake_case.

JinmingHu-MSFT commented:
@vhvb1989 I've fixed everything in Storage Blobs Service, @katmsft will fix DataLake service. However storage unittests still cannot pass. Some of the bugs in core code haven't been fixed yet, and some new ones are introduced in this PR.

@katmsft I disabled datalake build. https://github.com/Azure/azure-sdk-for-cpp/blob/74c84147c5b8d673ecddb58926704ae92ed49ea5/sdk/storage/CMakeLists.txt#L78 https://github.com/Azure/azure-sdk-for-cpp/blob/74c84147c5b8d673ecddb58926704ae92ed49ea5/sdk/storage/CMakeLists.txt#L83  https://github.com/Azure/azure-sdk-for-cpp/blob/74c84147c5b8d673ecddb58926704ae92ed49ea5/sdk/storage/sample/CMakeLists.txt#L11

katmsft commented on sdk/core/azure-core/src/http/bodyStream.cpp:
```diff
@@ -0,0 +1,126 @@
+
+std::unique_ptr<std::vector<uint8_t>> BodyStream::ReadToEnd(Context& context, BodyStream& body)
+{
+  constexpr int64_t chunkSize = 1024 * 8;
```
Magic numbers should be documented. Ideally, this is a changeable size for the client.

JinmingHu-MSFT replied:
@katmsft Let's be less picky about these trivial stuff and focus on the ability to function properly at this stage.

JinmingHu-MSFT commented:
@vhvb1989 There's only 1 bug left unfixed in https://github.com/Azure/azure-sdk-for-cpp/issues/198. After fixing that one, blob unit tests can pass.

Although I did see some flaky crash, but i'm not going to make big deal out of it. We can fix that later.

vhvb1989 commented:
storage unit tests passing:
![image](https://user-images.githubusercontent.com/24213737/86498079-21701300-bd39-11ea-97bf-892a6c854f2e.png)

@JinmingHu-MSFT , please approve changes to merge this PR.

katmsft commented on sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:
```diff
@@ -1063,8 +1063,7 @@ namespace Azure { namespace Storage { namespace Blobs {
           std::vector<uint8_t> bodyContent;
           if (bodyStream->Length() == -1)
```
What does this mean?

JinmingHu-MSFT replied:
unknown length, it's Azure Core's design.
